File: src/main/java/com/codingforcookies/betterrecords/client/models/ModelFrequencyTuner.java
package com.codingforcookies.betterrecords.client.models;

import java.awt.Color;

import org.lwjgl.opengl.GL11;

import net.minecraft.client.model.ModelBase;
import net.minecraft.client.model.ModelRenderer;
import net.minecraft.entity.Entity;
import net.minecraft.item.ItemStack;

public class ModelFrequencyTuner extends ModelBase {
	ModelRenderer Crystal_1;
	ModelRenderer Crystal_2;
	ModelRenderer Crystal_3;
	ModelRenderer Crystal_4;
	ModelRenderer Base;
	ModelRenderer Platform;
	ModelRenderer Stand;
	ModelRenderer TunerBridge;
	ModelRenderer TunerWeight;
	ModelRenderer Tuner;

	public ModelFrequencyTuner() {
		textureWidth = 64;
		textureHeight = 32;
		
		Crystal_1 = new ModelRenderer(this, 1, 0);
		Crystal_1.setTextureSize(64, 32);
		Crystal_1.addBox(-1F, -1F, -1F, 2, 2, 2);
		Crystal_1.setRotationPoint(0F, 18F, -3F);
		setRotation(Crystal_1, 0F, 0F, 0F);

		Crystal_2 = new ModelRenderer(this, 1, 0);
		Crystal_2.setTextureSize(64, 32);
		Crystal_2.addBox(-1F, -1F, -1F, 2, 2, 2);
		Crystal_2.setRotationPoint(0F, 18F, -3F);
		setRotation(Crystal_2, 0F, -0.7853982F, 0F);

		Crystal_3 = new ModelRenderer(this, 1, 0);
		Crystal_3.setTextureSize(64, 32);
		Crystal_3.addBox(-1F, -1F, -1F, 2, 2, 2);
		Crystal_3.setRotationPoint(0F, 18F, -3F);
		setRotation(Crystal_3, 0F, 0F, 0F);

		Crystal_4 = new ModelRenderer(this, 1, 0);
		Crystal_4.setTextureSize(64, 32);
		Crystal_4.addBox(-1F, -1F, -1F, 2, 2, 2);
		Crystal_4.setRotationPoint(0F, 18F, -3F);
		setRotation(Crystal_4, 0F, 0F, 0F);

		Base = new ModelRenderer(this, 0, 0);
		Base.addBox(-5F, 0F, -6F, 10, 9, 12);
		Base.setRotationPoint(0F, 15F, 0F);
		Base.setTextureSize(64, 32);
		Base.mirror = true;
		setRotation(Base, 0F, 0F, 0F);
		
		Platform = new ModelRenderer(this, 0, 21);
		Platform.addBox(-4F, -1F, -5F, 8, 1, 8);
		Platform.setRotationPoint(0F, 15F, 0F);
		Platform.setTextureSize(64, 32);
		Platform.mirror = true;
		setRotation(Platform, 0F, 0F, 0F);
		
		Stand = new ModelRenderer(this, 44, 0);
		Stand.addBox(-0.5F, -2.5F, 4F, 1, 3, 1);
		Stand.setRotationPoint(0F, 15F, 0F);
		Stand.setTextureSize(64, 32);
		Stand.mirror = true;
		setRotation(Stand, 0F, 0F, 0F);
		
		TunerBridge = new ModelRenderer(this, 44, 4);
		TunerBridge.addBox(-0.5F, -1F, 4F, 1, 6, 1);
		TunerBridge.setRotationPoint(0F, 15F, 0F);
		TunerBridge.setTextureSize(64, 32);
		TunerBridge.mirror = true;
		setRotation(TunerBridge, 1.041001F, 0F, 0F);
		
		TunerWeight = new ModelRenderer(this, 48, 0);
		TunerWeight.addBox(-1.5F, -2F, 5.5F, 3, 1, 2);
		TunerWeight.setRotationPoint(0F, 15F, 0F);
		TunerWeight.setTextureSize(64, 32);
		TunerWeight.mirror = true;
		setRotation(TunerWeight, 0F, 0F, 0F);
		
		Tuner = new ModelRenderer(this, 48, 3);
		Tuner.addBox(-0.5F, -2F, 3F, 1, 1, 2);
		Tuner.setRotationPoint(0F, 15F, 0F);
		Tuner.setTextureSize(64, 32);
		Tuner.mirror = true;
		setRotation(Tuner, 1.029744F, 0F, 0F);
	}
	
	private void setRotation(ModelRenderer model, float x, float y, float z) {
		model.rotateAngleX = x;
		model.rotateAngleY = y;
		model.rotateAngleZ = z;
	}

	public void render(Entity entity, float f, float f1, float f2, float f3, float f4, float f5, ItemStack crystal) {
		super.render(entity, f, f1, f2, f3, f4, f5);
		setRotationAngles(f, f1, f2, f3, f4, f5, entity);
		
		if(crystal != null) {
			GL11.glPushMatrix();
			{
				GL11.glEnable(GL11.GL_BLEND);
				if(crystal.stackTagCompound != null && crystal.stackTagCompound.hasKey("color")) {
					Color color = new Color(crystal.stackTagCompound.getInteger("color"));
					GL11.glColor4f(color.getRed() / 255F, color.getGreen() / 255F, color.getBlue() / 255F, .6F);
				}else
					GL11.glColor4f(1F, 1F, 1F, .6F);
				GL11.glScalef(.7F, .7F, .7F);
				GL11.glTranslatef(0F, .05F, -.1F);
				GL11.glRotatef(f, 0F, 1F, 0F);
				GL11.glTranslatef(0F, 0F, .18F);
				Crystal_1.render(f5);
				Crystal_2.render(f5);
				Crystal_3.render(f5);
				Crystal_4.render(f5);
				GL11.glColor3d(1F, 1F, 1F);
				GL11.glDisable(GL11.GL_BLEND);
			}
			GL11.glPopMatrix();
		}
		
		Base.render(f5);
		Platform.render(f5);
		Stand.render(f5);
		TunerBridge.render(f5);
		TunerWeight.render(f5);
		Tuner.render(f5);
	}
}

File: src/main/kotlin/com/codingforcookies/betterrecords/client/model/ModelFrequencyTuner.kt
package com.codingforcookies.betterrecords.client.model

import net.minecraft.client.model.ModelBase
import net.minecraft.client.model.ModelRenderer
import net.minecraft.client.renderer.GlStateManager
import net.minecraft.entity.Entity
import net.minecraft.item.ItemStack
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly
import java.awt.Color

@SideOnly(Side.CLIENT)
class ModelFrequencyTuner : ModelBase() {
    var Stand: ModelRenderer
    var Crystal_3: ModelRenderer
    var Platform: ModelRenderer
    var TunerBridge: ModelRenderer
    var TunerWeight: ModelRenderer
    var Crystal_4: ModelRenderer
    var Crystal_1: ModelRenderer
    var Crystal_2: ModelRenderer
    var Tuner: ModelRenderer
    var Base: ModelRenderer

    init {
        textureWidth = 64
        textureHeight = 32

        this.Stand = ModelRenderer(this, 44, 0).apply {
            setRotationPoint(0.0F, 15.0F, 0.0F)
            addBox(-0.5F, -2.5F, 4.0F, 1, 3, 1)
        }
        this.Crystal_3 = ModelRenderer(this, 1, 0).apply {
            setRotationPoint(0.0F, 18.0F, -3.0F)
            addBox(-1.0F, -1.0F, -1.0F, 2, 2, 2)
        }
        this.Platform = ModelRenderer(this, 0, 21).apply {
            setRotationPoint(0.0F, 15.0F, 0.0F)
            addBox(-4.0F, -1.0F, -5.0F, 8, 1, 8)
        }
        this.TunerBridge = ModelRenderer(this, 44, 4).apply {
            setRotationPoint(0.0F, 15.0F, 0.0F)
            addBox(-0.5F, -1.0F, 4.0F, 1, 6, 1)
        }
        this.setRotationAngles(this.TunerBridge, 1.0410009745512285F, 0.0F, 0.0F)
        this.TunerWeight = ModelRenderer(this, 48, 0).apply {
            setRotationPoint(0.0F, 15.0F, 0.0F)
            addBox(-1.5F, -2.0F, 5.5F, 3, 1, 2)
        }
        this.Crystal_4 = ModelRenderer(this, 1, 0).apply {
            setRotationPoint(0.0F, 18.0F, -3.0F)
            addBox(-1.0F, -1.0F, -1.0F, 2, 2, 2)
        }
        this.Crystal_1 = ModelRenderer(this, 1, 0).apply {
            setRotationPoint(0.0F, 18.0F, -3.0F)
            addBox(-1.0F, -1.0F, -1.0F, 2, 2, 2)
        }
        this.Crystal_2 = ModelRenderer(this, 1, 0).apply {
            setRotationPoint(0.0F, 18.0F, -3.0F)
            addBox(-1.0F, -1.0F, -1.0F, 2, 2, 2)
        }
        this.setRotationAngles(this.Crystal_2, 0.0F, -0.7853981633974483F, 0.0F)
        this.Tuner = ModelRenderer(this, 48, 3).apply {
            setRotationPoint(0.0F, 15.0F, 0.0F)
            addBox(-0.5F, -2.0F, 3.0F, 1, 1, 2)
        }
        this.setRotationAngles(this.Tuner, 1.0297440589395728F, 0.0F, 0.0F)
        this.Base = ModelRenderer(this, 0, 0).apply {
            setRotationPoint(0.0F, 15.0F, 0.0F)
            addBox(-5.0F, 0.0F, -6.0F, 10, 9, 12)
        }
    }

    fun render(entity: Entity?, f: Float, f1: Float, f2: Float, f3: Float, f4: Float, f5: Float, crystal: ItemStack) {
        super.render(entity, f, f1, f2, f3, f4, f5)
        setRotationAngles(f, f1, f2, f3, f4, f5, entity)

       if (!crystal?.isEmpty) {
            GlStateManager.pushMatrix()

            GlStateManager.enableBlend()

            if (crystal.hasTagCompound() && crystal.tagCompound!!.hasKey("color")) {
                val color = Color(crystal.tagCompound!!.getInteger("color"))
                GlStateManager.color(color.red / 255f, color.green / 255f, color.blue / 255f, .6f)
            } else
                GlStateManager.color(1f, 1f, 1f, .6f)

            GlStateManager.scale(.7f, .7f, .7f)
            GlStateManager.translate(0f, .05f, -.1f)
            GlStateManager.rotate(f, 0f, 1f, 0f)
            GlStateManager.translate(0f, 0f, .18f)

            Crystal_1.render(f5)
            Crystal_2.render(f5)
            Crystal_3.render(f5)
            Crystal_4.render(f5)

            GlStateManager.color(1.0f, 1.0f, 1.0f)
            GlStateManager.disableBlend()

            GlStateManager.popMatrix()
        }

        Base.render(f5)
        Platform.render(f5)
        Stand.render(f5)
        TunerBridge.render(f5)
        TunerWeight.render(f5)
        Tuner.render(f5)
    }

    private fun setRotationAngles(model: ModelRenderer, x: Float, y: Float, z: Float) {
        model.rotateAngleX = x
        model.rotateAngleY = y
        model.rotateAngleZ = z
    }
}
